File: src/lib/constraintValidation.ts
/**
 * Runtime validators for compile-time constraints (age vincoli).
 */
import type { CompiledAgeConstraint } from './agentBundleTypes';

/** Inclusive age check against compiled min/max bounds. */
export function satisfiesAgeYears(
  age: number,
  min: number | null,
  max: number | null,
): boolean {
  if (!Number.isFinite(age) || age < 0) return false;
  if (min != null && age < min) return false;
  if (max != null && age > max) return false;
  return true;
}

/** True when every age constraint on the path accepts the given age. */
export function pathSatisfiesAgeConstraints(
  age: number,
  constraints: CompiledAgeConstraint[],
): boolean {
  const ageRules = constraints.filter((c) => c.kind === 'age_years');
  if (ageRules.length === 0) return true;
  return ageRules.every((rule) => satisfiesAgeYears(age, rule.min, rule.max));
}

/** Inclusive age check against compiled week bounds (canonical). */
export function satisfiesAgeTotalWeeks(
  totalWeeks: number,
  minWeeks: number | null,
  maxWeeks: number | null,
  minMonths: number | null,
  maxMonths: number | null,
  min: number | null,
  max: number | null,
): boolean {
  if (!Number.isFinite(totalWeeks) || totalWeeks < 0) return false;
  if (minWeeks != null || maxWeeks != null) {
    if (minWeeks != null && totalWeeks < minWeeks) return false;
    if (maxWeeks != null && totalWeeks > maxWeeks) return false;
    return true;
  }
  if (minMonths != null || maxMonths != null) {
    const totalMonths = Math.floor((totalWeeks * 12) / 52);
    if (minMonths != null && totalMonths < minMonths) return false;
    if (maxMonths != null && totalMonths > maxMonths) return false;
    return true;
  }
  if (min != null && totalWeeks < min * 52) return false;
  if (max != null && totalWeeks > max * 52 + 51) return false;
  return true;
}

/** Inclusive age check against compiled min/max bounds (years or total months). */
export function satisfiesAgeTotalMonths(
  totalMonths: number,
  min: number | null,
  max: number | null,
  minMonths: number | null,
  maxMonths: number | null,
  minWeeks: number | null = null,
  maxWeeks: number | null = null,
): boolean {
  if (!Number.isFinite(totalMonths) || totalMonths < 0) return false;
  if (minWeeks != null || maxWeeks != null) {
    const totalWeeks = Math.floor((totalMonths * 52) / 12);
    return satisfiesAgeTotalWeeks(totalWeeks, minWeeks, maxWeeks, minMonths, maxMonths, min, max);
  }
  if (minMonths != null || maxMonths != null) {
    if (minMonths != null && totalMonths < minMonths) return false;
    if (maxMonths != null && totalMonths > maxMonths) return false;
    return true;
  }
  if (min != null && totalMonths < min * 12) return false;
  if (max != null && totalMonths > max * 12 + 11) return false;
  return true;
}

/** True when every age constraint on the path accepts the given total weeks. */
export function pathSatisfiesAgeConstraintsFromTotalWeeks(
  totalWeeks: number,
  constraints: CompiledAgeConstraint[],
): boolean {
  const ageRules = constraints.filter((c) => c.kind === 'age_years');
  if (ageRules.length === 0) return true;
  return ageRules.every((rule) =>
    satisfiesAgeTotalWeeks(
      totalWeeks,
      rule.minWeeks,
      rule.maxWeeks,
      rule.minMonths,
      rule.maxMonths,
      rule.min,
      rule.max,
    ),
  );
}

/** True when every age constraint on the path accepts the given total months. */
export function pathSatisfiesAgeConstraintsFromTotalMonths(
  totalMonths: number,
  constraints: CompiledAgeConstraint[],
): boolean {
  const ageRules = constraints.filter((c) => c.kind === 'age_years');
  if (ageRules.length === 0) return true;
  return ageRules.every((rule) =>
    satisfiesAgeTotalMonths(
      totalMonths,
      rule.min,
      rule.max,
      rule.minMonths,
      rule.maxMonths,
      rule.minWeeks,
      rule.maxWeeks,
    ),
  );
}

/** Italian cardinal words commonly spoken as patient age (voice input). */
const ITALIAN_AGE_WORDS: Readonly<Record<string, number>> = {
  zero: 0,
  uno: 1, una: 1,
  due: 2,
  tre: 3,
  quattro: 4,
  cinque: 5,
  sei: 6,
  sette: 7,
  otto: 8,
  nove: 9,
  dieci: 10,
  undici: 11,
  dodici: 12,
  tredici: 13,
  quattordici: 14,
  quindici: 15,
  sedici: 16,
  diciassette: 17,
  diciotto: 18,
  diciannove: 19,
  venti: 20,
  trenta: 30,
  trent: 30,
  quaranta: 40,
  cinquanta: 50,
  sessanta: 60,
  settanta: 70,
  ottanta: 80,
  novanta: 90,
  cento: 100,
};

function normalizeItalianAgeWord(raw: string): string {
  return raw
    .toLowerCase()
    .normalize('NFD')
    .replace(/\p{M}/gu, '')
    .replace(/'/g, '');
}

function parseItalianAgeWord(raw: string): number | null {
  const age = ITALIAN_AGE_WORDS[normalizeItalianAgeWord(raw)];
  return age != null && age >= 0 && age <= 120 ? age : null;
}

/** Extracts first plausible patient age in years from Italian free text. */
export function extractAgeYearsFromText(text: string): number | null {
  const normalized = text.trim().toLowerCase();
  if (!normalized) return null;

  const explicit = normalized.match(
    /(?:ho|ha|sono|è|e|di)\s*(\d{1,3})\s*anni?/,
  );
  if (explicit) {
    const age = Number.parseInt(explicit[1]!, 10);
    return Number.isFinite(age) ? age : null;
  }

  const bare = normalized.match(/\b(\d{1,3})\s*anni?\b/);
  if (bare) {
    const age = Number.parseInt(bare[1]!, 10);
    return Number.isFinite(age) ? age : null;
  }

  const wordWithVerb = normalized.match(
    /(?:ho|ha|sono|è|e|di)\s+([a-zàèéìòù']+)(?:\s+anni|\s*'anni)/,
  );
  if (wordWithVerb) {
    const age = parseItalianAgeWord(wordWithVerb[1]!);
    if (age != null) return age;
  }

  const wordWithAnni = normalized.match(/^([a-zàèéìòù']+)(?:\s+anni|\s*'anni)$/);
  if (wordWithAnni) {
    const age = parseItalianAgeWord(wordWithAnni[1]!);
    if (age != null) return age;
  }

  const standaloneWord = parseItalianAgeWord(normalized);
  if (standaloneWord != null) return standaloneWord;

  return null;
}

/** Standard voice question for age vincoli (never a fascia menu). */
export const AGE_YEARS_QUESTION = 'Quanti anni ha il paziente?';

File: dialog-engine/src/DialogEngine/ConstraintValidation.vb
''' <summary>
''' Runtime validators for compile-time age constraints and Italian age extraction.
''' </summary>
Imports System.Text.RegularExpressions

Public Module ConstraintValidation

    Public Const AgeYearsQuestion As String = "Quanti anni ha il paziente?"

    Private ReadOnly ItalianAgeWords As Dictionary(Of String, Integer) = New Dictionary(Of String, Integer)(StringComparer.OrdinalIgnoreCase) From {
        {"zero", 0}, {"uno", 1}, {"una", 1}, {"due", 2}, {"tre", 3}, {"quattro", 4},
        {"cinque", 5}, {"sei", 6}, {"sette", 7}, {"otto", 8}, {"nove", 9}, {"dieci", 10},
        {"undici", 11}, {"dodici", 12}, {"tredici", 13}, {"quattordici", 14}, {"quindici", 15},
        {"sedici", 16}, {"diciassette", 17}, {"diciotto", 18}, {"diciannove", 19},
        {"venti", 20}, {"trenta", 30}, {"trent", 30}, {"quaranta", 40}, {"cinquanta", 50},
        {"sessanta", 60}, {"settanta", 70}, {"ottanta", 80}, {"novanta", 90}, {"cento", 100}
    }

    Public Function SatisfiesAgeYears(age As Integer, min As Integer?, max As Integer?) As Boolean
        If age < 0 Then Return False
        If min.HasValue AndAlso age < min.Value Then Return False
        If max.HasValue AndAlso age > max.Value Then Return False
        Return True
    End Function

    Public Function PathSatisfiesAgeConstraints(age As Integer, constraints As IList(Of Models.AgeConstraint)) As Boolean
        If constraints Is Nothing OrElse constraints.Count = 0 Then Return True
        Return constraints.All(Function(rule) SatisfiesAgeYears(age, rule.Min, rule.Max))
    End Function

    Public Function SatisfiesAgeTotalMonths(totalMonths As Integer, min As Integer?, max As Integer?) As Boolean
        If totalMonths < 0 Then Return False
        If min.HasValue AndAlso totalMonths < min.Value * 12 Then Return False
        If max.HasValue AndAlso totalMonths > (max.Value * 12) + 11 Then Return False
        Return True
    End Function

    Public Function SatisfiesAgeConstraintTotalWeeks(
        totalWeeks As Integer,
        rule As Models.AgeConstraint
    ) As Boolean
        If totalWeeks < 0 Then Return False
        If rule.MinWeeks.HasValue OrElse rule.MaxWeeks.HasValue Then
            If rule.MinWeeks.HasValue AndAlso totalWeeks < rule.MinWeeks.Value Then Return False
            If rule.MaxWeeks.HasValue AndAlso totalWeeks > rule.MaxWeeks.Value Then Return False
            Return True
        End If
        Dim totalMonths = (totalWeeks * 12 + 26) \ 52
        Return SatisfiesAgeConstraintTotalMonths(totalMonths, rule)
    End Function

    Public Function PathSatisfiesAgeConstraintsFromTotalWeeks(
        totalWeeks As Integer,
        constraints As IList(Of Models.AgeConstraint)
    ) As Boolean
        If constraints Is Nothing OrElse constraints.Count = 0 Then Return True
        Return constraints.All(Function(rule) SatisfiesAgeConstraintTotalWeeks(totalWeeks, rule))
    End Function

    Public Function SatisfiesAgeConstraintTotalMonths(
        totalMonths As Integer,
        rule As Models.AgeConstraint
    ) As Boolean
        If totalMonths < 0 Then Return False
        If rule.MinMonths.HasValue OrElse rule.MaxMonths.HasValue Then
            If rule.MinMonths.HasValue AndAlso totalMonths < rule.MinMonths.Value Then Return False
            If rule.MaxMonths.HasValue AndAlso totalMonths > rule.MaxMonths.Value Then Return False
            Return True
        End If
        Return SatisfiesAgeTotalMonths(totalMonths, rule.Min, rule.Max)
    End Function

    Public Function PathSatisfiesAgeConstraintsFromTotalMonths(
        totalMonths As Integer,
        constraints As IList(Of Models.AgeConstraint)
    ) As Boolean
        If constraints Is Nothing OrElse constraints.Count = 0 Then Return True
        Return constraints.All(Function(rule) SatisfiesAgeConstraintTotalMonths(totalMonths, rule))
    End Function

    Public Function ExtractAgeYearsFromText(text As String) As Integer?
        If String.IsNullOrWhiteSpace(text) Then Return Nothing
        Dim normalized = AgeUtteranceNormalize.NormalizeAgeUtterance(text)
        If String.IsNullOrWhiteSpace(normalized) Then Return Nothing

        Dim explicitMatch = Regex.Match(normalized, "(?:ho|ha|sono|è|e|di)\s*(\d{1,3})\s*anni?")
        If explicitMatch.Success Then
            Return Integer.Parse(explicitMatch.Groups(1).Value)
        End If

        Dim bareMatch = Regex.Match(normalized, "\b(\d{1,3})\s*anni?\b")
        If bareMatch.Success Then
            Return Integer.Parse(bareMatch.Groups(1).Value)
        End If

        Dim wordWithVerb = Regex.Match(normalized, "(?:ho|ha|sono|è|e|di)\s+([a-zàèéìòù']+)(?:\s+anni|\s*'anni)")
        If wordWithVerb.Success Then
            Dim age = ParseItalianAgeWord(wordWithVerb.Groups(1).Value)
            If age.HasValue Then Return age
        End If

        Dim wordWithAnni = Regex.Match(normalized, "^([a-zàèéìòù']+)(?:\s+anni|\s*'anni)$")
        If wordWithAnni.Success Then
            Dim age = ParseItalianAgeWord(wordWithAnni.Groups(1).Value)
            If age.HasValue Then Return age
        End If

        If Regex.IsMatch(normalized, "^\d{1,3}$") Then
            Dim bareAge = Integer.Parse(normalized)
            If bareAge >= 0 AndAlso bareAge <= 120 Then Return bareAge
        End If

        Return ParseItalianAgeWord(normalized)
    End Function

    Private Function ParseItalianAgeWord(raw As String) As Integer?
        Dim key = CategoryNormalization.NormalizeWordKey(raw)
        If ItalianAgeWords.TryGetValue(key, Nothing) Then
            Dim age = ItalianAgeWords(key)
            If age >= 0 AndAlso age <= 120 Then Return age
        End If
        Return Nothing
    End Function

End Module
